### ISSUE TITLE
hip::device in C/C++/Fortran Codes

### ISSUE THREAD

**Issue by ax3l** [ROCm-Developer-Tools/HIP#2275]
Hi,

I am trying to modernize the CMake logic of [ECP AMReX](https://github.com/AMReX-Codes/amrex/) and [ECP WarpX](https://github.com/ECP-WarpX/WarpX) away from `hipcc` to the recommended CMake logic (due to limitations of the `hipcc` perl script).

Following the guidance in the [docs](https://rocmdocs.amd.com/en/latest/Installation_Guide/Using-CMake-with-AMD-ROCm.html#using-hip-in-cmake), I link the `hip::device` target in my projects.

It seems the flags it sets are unfortunately not limited to the C/C++ language but also apply to the `Fortran` objects in our project (e.g. Fortran bindings). While this worked through `hipcc`, the `hip::device` target thus fails to compile.

An alternative approach to use the shipped `flang` executable for Fortran bindings is not recognized as compiler by the shipped CMake version. Generally, `flang` is not yet at the quality level (code gen.) of other Fortran compilers and should thus not be enforced (for host-only parts of a CMake project).

I think what probably needs to be done is to limit the compiler & linker flags of the `hip::device` object to C/C++ objects with generator expressions.

Xref: likely fix in #2190 is approved but unmerged since Nov 2020
Xref: partly related to #2158
Xref: comment https://github.com/ROCm-Developer-Tools/HIP/issues/2158#issuecomment-727232969
Xref: comment https://github.com/ROCm-Developer-Tools/HIP/issues/2154#issuecomment-843690799

**Comment by ax3l** [comment]
Does this repo change the default of `-std=c++<...>` for `clang++`?
I am trying to transition to this logic, the compiler identifies as Clang 12 but it builds by default in C++11.

My targets set
```cmake
target_compile_features(${tgt} PUBLIC cxx_std_14)
```

and only with this `clang++` in hip 4.1.0 I see no flags getting added to the command line.

Note that Clang 6+ already default to C++14: https://gist.github.com/ax3l/53db9fa8a4f4c21ecc5c4100c0d93c94

But `-xhip` does not it seems:
```
rocm-4.1.0/llvm/bin/clang++ -dM -E -xhip  /dev/null | grep -F __cplusplus
#define __cplusplus 201103L
#define __cplusplus 201103L
```

**Comment by ax3l** [comment]
Fix ready for merge in #2280

**Comment by ax3l** [comment]
Problem solved at least with ROCm 4.5.0 :tada:
